spark3/src/main/java/org/apache/iceberg/spark/SparkCatalog.java

```diff
@@ -0,0 +1,280 @@
+        propertyChanges.add(change);
+      } else {
+        schemaChanges.add(change);
+      }
+    }
+
+    try {
+      Table table = icebergCatalog.loadTable(buildIdentifier(ident));
+      commitChanges(table, setLocation, setSnapshotId, pickSnapshotId, propertyChanges, schemaChanges);
+    } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
+      throw new NoSuchTableException(ident);
+    }
+
+    return null;
+  }
+
+  @Override
+  public boolean dropTable(Identifier ident) {
+    try {
+      return icebergCatalog.dropTable(buildIdentifier(ident), true);
+    } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
+      return false;
+    }
+  }
+
+  @Override
+  public void renameTable(Identifier from, Identifier to) throws NoSuchTableException, TableAlreadyExistsException {
+    try {
+      icebergCatalog.renameTable(buildIdentifier(from), buildIdentifier(to));
+    } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
+      throw new NoSuchTableException(from);
+    } catch (AlreadyExistsException e) {
+      throw new TableAlreadyExistsException(to);
+    }
+  }
+
+  @Override
+  public void invalidateTable(Identifier ident) {
+    try {
+      icebergCatalog.loadTable(buildIdentifier(ident)).refresh();
+    } catch (org.apache.iceberg.exceptions.NoSuchTableException ignored) {
+      // ignore if the table doesn't exist, it is not cached
+    }
+  }
+
+  @Override
+  public final void initialize(String name, CaseInsensitiveStringMap options) {
+    boolean cacheEnabled = Boolean.parseBoolean(options.getOrDefault("cache-enabled", "true"));
+    Catalog catalog = buildIcebergCatalog(name, options);
+
+    this.catalogName = name;
+    this.icebergCatalog = cacheEnabled ? CachingCatalog.wrap(catalog) : catalog;
+  }
+
+  @Override
+  public String name() {
+    return catalogName;
+  }
+
+  private static void commitChanges(Table table, SetProperty setLocation, SetProperty setSnapshotId,
+                                    SetProperty pickSnapshotId, List<TableChange> propertyChanges,
+                                    List<TableChange> schemaChanges) {
+    // don't allow setting the snapshot adn picking a commit at the same time because order is ambiguous and choosing
+    // one order leads to different results
+    Preconditions.checkArgument(setSnapshotId == null || pickSnapshotId == null,
+        "Cannot set the current the current snapshot ID and cherry-pick snapshot changes");
+
+    if (setSnapshotId != null) {
+      long newSnapshotId = Long.parseLong(setSnapshotId.value());
+      table.manageSnapshots().setCurrentSnapshot(newSnapshotId).commit();
+    }
+
+    // if updating the table snapshot, perform that update first in case it fails
+    if (pickSnapshotId != null) {
+      long newSnapshotId = Long.parseLong(pickSnapshotId.value());
+      table.manageSnapshots().cherrypick(newSnapshotId).commit();
+    }
+
+    Transaction transaction = table.newTransaction();
```

Comment:
In which cases will we both update props and modify schema?

Reply:
I originally proposed a public API for updating tables, which could batch operations together into a group:

```java
Table updated = spark.catalog("prod")
    .updateTable("name")
    .addColumn("z", "bigint")
    .renameColumn("a", "x")
    .set("prop", "value")
```

I think it makes sense to assume that this may happen. We could also receive direct calls through the catalog API.